### ISSUE TITLE
Basic testing errors, stack not reachable or usable

### ISSUE THREAD

**Issue by stobbsm** [CanDIG/CanDIGv2#373]
I've run clean from both dev and stable branchs, same result no matter what.

I took the extra steps of loading in the ingestion test data (which completes successfully surprisingly), but end up with the same errors regardless.

Not sure where to look at this point, but I'm going through the `Makefile` to see if I can start tracking down a cause.

Running on virtualized Rocky linux 9.3 with 128G of memory, 8 cores of Icelake xeon

Please add the results of the following commands:

```
git status
On branch stable
Your branch is up to date with 'origin/stable'.

nothing to commit, working tree clean
```

```
git submodule status
 fb9c83a323d8624495c5233a5afdd7b25f869fc0 lib/candig-data-portal/candig-data-portal (v1.1.1)
 18947077842ce0bc442735c62442519ace18616d lib/candig-ingest/candigv2-ingest (v1.1.0-8-g1894707)
 9303e0b310c1cf2419f8796bde22717ab0807223 lib/federation/federation (v1.1.0-1-g9303e0b)
 745f0b1800e940276e361227f334432c570fa38a lib/htsget/htsget_app (v2.2.0-58-g745f0b1)
 01f879bf6788b729faca0d96a7b6144ba2a5eaee lib/katsu/katsu_service (v2.3.0-7-g01f879bf)
 4f2684ee1c3b35ff6b250651cd6fdc8bd77a9a77 lib/opa/opa (v1.3.3-6-g4f2684e)
 f12540a3f79936134eec2aeb356da98fcbb526e5 lib/query/query (v1.0.0-8-gf12540a)
 b0ff5be051f2fd55352e00450b7848dcf8354a3b lib/toil/toil-docker (releases/5.5.0)
```

```
cat tmp/error.txt
[error.txt](https://github.com/CanDIG/CanDIGv2/files/13833527/error.txt)

```

```
make test-integration
.env:124: extraneous text after 'define' directive
python ./settings.py
source ./env.sh; pytest ./etc/tests
============================================================================================================= test session starts ==============================================================================================================
platform linux -- Python 3.10.13, pytest-7.2.0, pluggy-1.3.0
rootdir: /home/rocky/CanDIGv2
plugins: anyio-4.2.0
collected 25 items                                                                                                                                                                                                                             

etc/tests/test_integration.py .....................FFFF                                                                                                                                                                                  [100%]

=================================================================================================================== FAILURES ===================================================================================================================
______________________________________________________________________________________________________________ test_server_count _______________________________________________________________________________________________________________

    def test_server_count():
        token = get_token(
            username=ENV["CANDIG_NOT_ADMIN_USER"], password=ENV["CANDIG_NOT_ADMIN_PASSWORD"]
        )
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json; charset=utf-8",
        }
        response = requests.get(
            f"{ENV['CANDIG_URL']}/federation/v1/servers", headers=headers
        )
        print(response.json())
>       assert len(response.json()) > 0
E       assert 0 > 0
E        +  where 0 = len([])
E        +    where [] = <bound method Response.json of <Response [200]>>()
E        +      where <bound method Response.json of <Response [200]>> = <Response [200]>.json

etc/tests/test_integration.py:861: AssertionError
------------------------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------------------------
[]
_____________________________________________________________________________________________________________ test_services_count ______________________________________________________________________________________________________________

    def test_services_count():
        token = get_token(
            username=ENV["CANDIG_NOT_ADMIN_USER"], password=ENV["CANDIG_NOT_ADMIN_PASSWORD"]
        )
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json; charset=utf-8",
        }
        response = requests.get(
            f"{ENV['CANDIG_URL']}/federation/v1/services", headers=headers
        )
        print(response.json())
>       assert len(response.json()) > 0
E       assert 0 > 0
E        +  where 0 = len([])
E        +    where [] = <bound method Response.json of <Response [200]>>()
E        +      where <bound method Response.json of <Response [200]>> = <Response [200]>.json

etc/tests/test_integration.py:877: AssertionError
------------------------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------------------------
[]
_____________________________________________________________________________________________________________ test_federation_call _____________________________________________________________________________________________________________

    def test_federation_call():
        body = {
            "service": "htsget",
            "method": "GET",
            "payload": {},
            "path": "beacon/v2/service-info",
        }
    
        token = get_token(
            username=ENV["CANDIG_SITE_ADMIN_USER"],
            password=ENV["CANDIG_SITE_ADMIN_PASSWORD"],
        )
        headers = {
            "Authorization": f"Bearer {token}",
            "content-type": "application/json",
            "federation": "false",
        }
    
        response = requests.post(
            f"{ENV['CANDIG_URL']}/federation/v1/fanout", headers=headers, json=body
        )
        print(response.json())
        assert "results" in response.json()
    
        headers["federation"] = "true"
        response = requests.post(
            f"{ENV['CANDIG_URL']}/federation/v1/fanout", headers=headers, json=body
        )
        print(response.json())
        assert "list" in str(type(response.json()))
>       assert "results" in response.json()[0]
E       IndexError: list index out of range

etc/tests/test_integration.py:913: IndexError
------------------------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------------------------
{'results': {}, 'service': 'htsget', 'status': 500}
[]
_______________________________________________________________________________________________________________ test_add_server ________________________________________________________________________________________________________________

    def test_add_server():
        token = get_token(
            username=ENV["CANDIG_SITE_ADMIN_USER"],
            password=ENV["CANDIG_SITE_ADMIN_PASSWORD"],
        )
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json; charset=utf-8",
        }
    
        response = requests.get(
            f"{ENV['CANDIG_URL']}/federation/v1/servers", headers=headers
        )
    
        body = {
>           "server": response.json()[0],
            "authentication": {"issuer": ENV["KEYCLOAK_REALM_URL"], "token": token},
        }
E       IndexError: list index out of range

etc/tests/test_integration.py:932: IndexError
=========================================================================================================== short test summary info ============================================================================================================
FAILED etc/tests/test_integration.py::test_server_count - assert 0 > 0
FAILED etc/tests/test_integration.py::test_services_count - assert 0 > 0
FAILED etc/tests/test_integration.py::test_federation_call - IndexError: list index out of range
FAILED etc/tests/test_integration.py::test_add_server - IndexError: list index out of range
========================================================================================================= 4 failed, 21 passed in 6.71s =========================================================================================================
make: *** [Makefile:521: test-integration] Error 1
```

**Comment by OrdiNeu** [comment]
It looks like federation has failed to initialize (the first two tests show zero servers and zero services available), which causes the rest of the errors. Federation handles the communications between several services, so an error there causes the rest of the integration tests to fail (the tests before that point all communicate directly with whichever microservice).

When taking federatiion up/down, do any errors appear in the log? The build process throws a lot of messages at the user, so error messages often get lost in the logs. For reference, to bring just federation down&up, run the following:

```make clean-federation; make build-federation; make compose-federation;```

**Comment by daisieh** [comment]
The end of `make compose-federation` should look like this:

```
2024-01-06T05:41:50.5505628Z Adding servers to federation...
2024-01-06T05:41:50.5506313Z POST response: 201 {
2024-01-06T05:41:50.5507017Z   "id": "internal-1",
2024-01-06T05:41:50.5507451Z   "location": {
2024-01-06T05:41:50.5507866Z     "name": "Local",
2024-01-06T05:41:50.5508326Z     "province": "ON",
2024-01-06T05:41:50.5508852Z     "province-code": "ca-on"
2024-01-06T05:41:50.5509303Z   },
2024-01-06T05:41:50.5509814Z   "url": "http://candig.docker.internal:4232/federation"
2024-01-06T05:41:50.5513188Z }
2024-01-06T05:41:50.5513416Z 
2024-01-06T05:41:50.5513570Z [
2024-01-06T05:41:50.5513873Z   {
2024-01-06T05:41:50.5514285Z     "id": "internal-1",
2024-01-06T05:41:50.5514726Z     "location": {
2024-01-06T05:41:50.5515121Z       "name": "Local",
2024-01-06T05:41:50.5515405Z       "province": "ON",
2024-01-06T05:41:50.5515722Z       "province-code": "ca-on"
2024-01-06T05:41:50.5515992Z     },
2024-01-06T05:41:50.5516295Z     "url": "http://candig.docker.internal:4232/federation"
2024-01-06T05:41:50.5516685Z   }
2024-01-06T05:41:50.5516881Z ]
2024-01-06T05:41:50.5516989Z 
2024-01-06T05:41:50.5517107Z Adding services to federation...
2024-01-06T05:41:50.5517397Z [
2024-01-06T05:41:50.5517589Z   {
2024-01-06T05:41:50.5517785Z     "id": "katsu",
2024-01-06T05:41:50.5518101Z     "url": "http://candig.docker.internal:8008",
2024-01-06T05:41:50.5518467Z     "version": "v2.3.0"
2024-01-06T05:41:50.5518715Z   },
2024-01-06T05:41:50.5518903Z   {
2024-01-06T05:41:50.5519098Z     "id": "htsget",
2024-01-06T05:41:50.5519391Z     "url": "http://candig.docker.internal:3333",
2024-01-06T05:41:50.5519747Z     "version": "v3.0.0"
2024-01-06T05:41:50.5519987Z   },
2024-01-06T05:41:50.5520175Z   {
2024-01-06T05:41:50.5520367Z     "id": "query",
2024-01-06T05:41:50.5520647Z     "url": "http://candig.docker.internal:1236",
2024-01-06T05:41:50.5520996Z     "version": "1.0.0"
2024-01-06T05:41:50.5521237Z   }
2024-01-06T05:41:50.5521421Z ]
2024-01-06T05:41:50.5521527Z 
2024-01-06T05:41:50.5686216Z echo "    finished compose-federation" >> tmp/progress.txt
```

**Comment by stobbsm** [comment]
Sorry, I got distracted.
At this point, rebuilding the federation stack with the `make` commands will get the stack up, but anytime I start the stack from scratch, I need to do the federation `clean`, `build` and `compose` step. Basically it fails every time from scratch.

This does unblock what I am doing atm though. I'll come back and try to figure out where the consistent error is when I'm able.

**Comment by mshadbolt** [comment]
Closing this as it is quite old and our stack has changed quite a bit, don't hesitate to create a new issue if you are still having troubles!
